File: csrc/joystick.c
#include "window-private.h"

#include "joystick.h"

Joystick** getJoysticks(Window* w) { return w->joysticks; }
unsigned int getJoystickId(Joystick* j) { return j->id; }
int getJoystickNumberOfAxii(Joystick* j) { return j->numberOfAxii; }
int getJoystickIsAdd(Joystick* j) { return j->isAdd; }

JoystickButton** getJoystickButtons(Window* w) { return w->joystickButtons; }
unsigned int joystickButtonId(JoystickButton* jb) { return jb->id; }
int joystickButtonIsPress(JoystickButton* jb) { return jb->isPress; }
unsigned int joystickButtonCode(JoystickButton* jb) { return jb->code; }


JoystickAxis** getJoystickAxii(Window* w) { return w->joystickAxii; }
unsigned int joystickAxisId(JoystickAxis* ja) { return ja->id; }
unsigned int joystickAxisAxis(JoystickAxis* ja) { return ja->axis; }
double joystickAxisValue(JoystickAxis* ja) { return ja->value; }

File: src/Graphics/Cogh/Joystick.hs
module Graphics.Cogh.Joystick
  ( maybeJoystickButton
  , maybeJoystickAxis
  ) where

import Graphics.Cogh.CommonFFI

maybeJoystickButton :: Window -> IO (Maybe Event)
maybeJoystickButton = maybeEvent getJoystickButton $ \ p -> do
  id' <- joystickButtonId p
  code <- joystickButtonCode p
  isPress <- joystickButtonIsPress p
  return $ JoystickButton (fromIntegral id')
    (fromIntegral code) (cBool isPress)

maybeJoystickAxis :: Window -> IO (Maybe Event)
maybeJoystickAxis = maybeEvent getJoystickAxis $ \ p -> do
  id' <- joystickAxisId p
  axis <- joystickAxisAxis p
  value <- joystickAxisValue p
  return $ JoystickAxis (fromIntegral id')
    (fromIntegral axis) (realToFrac value)


foreign import ccall unsafe "getJoystickButton" getJoystickButton
  :: Window -> IO (Ptr ())

foreign import ccall unsafe "joystickButtonId" joystickButtonId
  :: Ptr () -> IO (CInt)

foreign import ccall unsafe "joystickButtonIsPress" joystickButtonIsPress
  :: Ptr () -> IO (CInt)

foreign import ccall unsafe "joystickButtonCode" joystickButtonCode
  :: Ptr () -> IO (CUInt)


foreign import ccall unsafe "getJoystickAxis" getJoystickAxis
  :: Window -> IO (Ptr ())

foreign import ccall unsafe "joystickAxisId" joystickAxisId
  :: Ptr () -> IO (CInt)

foreign import ccall unsafe "joystickAxisAxis" joystickAxisAxis
  :: Ptr () -> IO (CUInt)

foreign import ccall unsafe "joystickAxisValue" joystickAxisValue
  :: Ptr () -> IO (CDouble)
